src/main/scala/is/hail/io/bgen/BgenWriter.scala

```diff
@@ -0,0 +1,343 @@
+      val r = if (indexInverse(i) < F) resized(i).ceil else resized(i).floor
+      if (keepLast || i != n - 1)
+        output += r.toUInt.intRep
+      newSize += r
+      i += 1
+    }
+    assert(newSize == expectedSize)
+  }
+
+  def headerBlock(sampleIds: Array[String], nVariants: Long): Array[Byte] = {
+    val bb = new ArrayBuilder[Byte]
+    emitHeaderBlock(bb, sampleIds, nVariants)
+    bb.result()
+  }
+
+  def emitHeaderBlock(bb: ArrayBuilder[Byte], sampleIds: Array[String], nVariants: Long) {
+    val nSamples = sampleIds.length
+    assert(nVariants <= (1L << 32) - 1, s"Maximum number of variants can export is (2^32 - 1). Found $nVariants.")
+
+    val magicNumbers = Array("b", "g", "e", "n").flatMap(_.getBytes)
+    val flags = 0x01 | (0x02 << 2) | (0x01 << 31)
+    val headerLength = 20
+
+    intToBytesLE(bb, 0) // placeholder for offset
+    intToBytesLE(bb, headerLength)
+    intToBytesLE(bb, nVariants.toUInt.intRep)
+    intToBytesLE(bb, nSamples)
+    bb ++= magicNumbers
+    intToBytesLE(bb, flags)
+
+    intToBytesLE(bb, 0) // placeholder for length of sample ids
+    intToBytesLE(bb, nSamples)
+
+    var i = 0
+    var sampleBlockLength = 8
+    while (i < nSamples) {
+      sampleBlockLength += stringToBytesWithShortLength(bb, sampleIds(i))
+      i += 1
+    }
+
+    val offset = headerLength + sampleBlockLength
+    updateIntToBytesLE(bb, offset, 0)
+    updateIntToBytesLE(bb, sampleBlockLength, 24)
+  }
+
+  def emitVariant(bb: ArrayBuilder[Byte], v: Variant, va: Annotation, view: ArrayGenotypeView,
+    rsidQuery: Option[Querier], varidQuery: Option[Querier], nSamples: Int, nBitsPerProb: Int) {
+    val nAlleles = v.nAlleles
+    require(nAlleles <= 0xffff, s"Maximum number of alleles per variant is ${ 0xffff }. Found ${ v.nAlleles }.")
+
+    val varid = varidQuery.flatMap(q => Option(q(va))).map(_.asInstanceOf[String]).getOrElse(v.toString)
+    val rsid = rsidQuery.flatMap(q => Option(q(va))).map(_.asInstanceOf[String]).getOrElse(".")
+
+    stringToBytesWithShortLength(bb, varid)
+    stringToBytesWithShortLength(bb, rsid)
+    stringToBytesWithShortLength(bb, v.contig)
+    intToBytesLE(bb, v.start)
+    shortToBytesLE(bb, nAlleles)
+    stringToBytesWithIntLength(bb, v.ref)
+    v.altAlleles.foreach(a => stringToBytesWithIntLength(bb, a.alt))
+
+    val gtDataBlockStart = bb.length
+    intToBytesLE(bb, 0) // placeholder for length of compressed data
+    intToBytesLE(bb, 0) // placeholder for length of uncompressed data
+
+    val uncompressedData = emitGPData(view, nSamples, nAlleles, nBitsPerProb)
+
+    val uncompressedLength = uncompressedData.length
+    val compressedLength = compress(bb, uncompressedData)
+
+    updateIntToBytesLE(bb, compressedLength + 4, gtDataBlockStart)
+    updateIntToBytesLE(bb, uncompressedLength, gtDataBlockStart + 4)
+  }
+
+  def emitGPData(view: ArrayGenotypeView, nSamples: Int, nAlleles: Int, nBitsPerProb: Int): Array[Byte] = {
+    val bb = new ArrayBuilder[Byte]
```

Comment:
This should be allocated once per partition.

I'd create a class to contain the per-partition state that you need to do everything.  This should include the overall bb, the uncompressed bb, and anything else.  The per-genotype arrays will still need to be allocated per-variant, of course.  Then you can move these routines onto that class.  This should simplify the interface for these routines because you won't have to pass all these values between the routines.  You can do the same thing for per-variant emission if needed (an object holding the nGenotype length arrays).